csi/node_server.go

```diff
@@ -690,11 +695,66 @@ func (ns *NodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVo
 	}, nil
 }
 
+// NodeExpandShared Volume is designed to expand the file system in an RWX volume for ONLINE expansion.
+// It does so with a gRPC call into the share-manager pod.
+func (ns *NodeServer) NodeExpandSharedVolume(volumeName string) error {
+	log := ns.log.WithFields(logrus.Fields{"function": "NodeExpandSharedVolume"})
+
+	lhNamespace := os.Getenv(types.EnvPodNamespace)
+	if lhNamespace == "" {
+		return fmt.Errorf("failed to detect pod namespace, environment variable %v is missing", types.EnvPodNamespace)
+	}
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return errors.Wrap(err, "failed to get client config")
+	}
+
+	kubeClient, err := clientset.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "failed to get k8s client")
+	}
+
+	lhClient, err := lhclientset.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "failed to get longhorn clientset")
+	}
+
+	sm, err := lhClient.LonghornV1beta2().ShareManagers(lhNamespace).Get(context.TODO(), volumeName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to get ShareManager CR")
+	}
+
+	podName := types.GetShareManagerPodNameFromShareManagerName(sm.Name)
+	pod, err := kubeClient.CoreV1().Pods(lhNamespace).Get(context.TODO(), podName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to get ShareManager pod")
+	}
+
+	client, err := engineapi.NewShareManagerClient(sm, pod)
+	if err != nil {
+		return errors.Wrapf(err, "failed to launch gRPC client for share manager before resizing volume %v", volumeName)
+	}
```

Comment:
I think we can get IP of share manager service by accessing `volume.ShareEndpoint` without having to look up the share manager pod?

Reply:
Yes, `volume.ShareEndpoint` is the IP of the service which should eventually forward to the pod

Reply:
Doesn't seem to work.  Here is a bit of the share-manager pod
```
 podIP: 10.42.215.169
```
And the volume
```
    shareEndpoint: nfs://10.43.184.173/pvc-eb61e66c-e54d-473d-8e14-b98c5bccae2f
```
Extracting the IP address and using it as the gRPC target,
```
[longhorn-csi-plugin-rmz9g longhorn-csi-plugin] W1127 21:26:31.201924       1 logging.go:55] [core] [Channel #119 SubChannel #120]grpc: addrConn.createTransport failed to connect to {Addr: "10.43.184.173:9600", ServerName: "10.43.184.173:9600", }. Err: connection error: desc = "transport: Error while dialing: dial tcp 10.43.184.173:9600: i/o timeout" 
[longhorn-csi-plugin-rmz9g longhorn-csi-plugin] time="2024-11-27T21:26:31Z" level=info msg="NodeExpandSharedVolume returned error." func="csi.(*NodeServer).NodeExpandVolume" file="node_server.go:806" component=csi-node-server error="rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial tcp 10.43.184.173:9600: i/o timeout\"" function=NodeExpandVolume 
[longhorn-csi-plugin-rmz9g longhorn-csi-plugin] time="2024-11-27T21:26:31Z" level=error msg="NodeExpandVolume: err: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial tcp 10.43.184.173:9600: i/o timeout\"" func=csi.logGRPC file="server.go:136" 
```

Reply:
It would be possible to patch the share-manager service that exports the NFS port to add the gRPC port as well.  But that would mean mixing data-plane and control plane.  It is preferred not to do that.  We will stick with the pod lookup, although it is requested to cache the kubeclient and lhclient.

Reply:
Done.